File: Qytetet Ruby/pkg/QytetetRuby-0.0.1/lib/vista_textual_qytetet.rb
# To change this license header, choose License Headers in Project Properties.
# To change this template file, choose Tools | Templates
# and open the template in the editor.
require_relative "controlador_qytetet"
require_relative "opcion_menu"
module VistaTextualQytetet
    class VistaTextualQytetet
      @@controlador=ControladorQytetet::ControladorQytetet.instance
      
      def obtener_nombre_jugadores
      puts "Introduzca el numero de jugadores"
      numero=gets.chomp.to_i
      nombres=Array.new
      
      for i in 1..numero
        puts "Introduzca el nombre del jugador #{i}"
        nombres.push(gets.chomp)
      end      
      return nombres
      end
      
      def elegir_casilla(opcion_menu)
        lista=@@controlador.obtener_casillas_validas(opcion_menu)
        casillas=Array.new
        if lista.empty?
          return -1
        else
          puts "Las casillas validas son: "
          for i in lista.each 
            puts i
            casillas<<i.to_s
          end
        end
        return leer_valor_correcto(casillas).to_i      
      end
      
      def leer_valor_correcto(valores_correctos)
        correcto=false
        valor_retornado =""
        puts "Selecciona una de las siguientes opciones: "
        
        for i in valores_correctos.each
          puts "Opcion: "+ i+" - " + ControladorQytetet::OpcionMenu[i.to_i].to_s
        end
        
        loop do
          valor_leido = gets.chomp.to_i
          for i in valores_correctos.each
            
          if valor_leido.to_i == i.to_i
            correcto=true
            valor_retornado = valor_leido
          end      
          
        end
        
        break if correcto
      end
      return valor_retornado
    end
    
    def elegir_operacion
      lista=@@controlador.obtener_operaciones_juego_validas()
      operaciones=Array.new
      for i in lista
        operaciones<<i.to_s
      end
      return leer_valor_correcto(operaciones).to_i
    end
    
    def self.main
      ui= VistaTextualQytetet.new
      @@controlador.set_nombre_jugadores(ui.obtener_nombre_jugadores)
      loop do
        operacion_elegida=ui.elegir_operacion
        necesita_elegir_casilla=@@controlador.necesita_elegir_casilla(operacion_elegida)
        if necesita_elegir_casilla
          casilla_elegida=ui.elegir_casilla(operacion_elegida)
        end
        if !necesita_elegir_casilla || casilla_elegida>=0
          puts @@controlador.realizar_operacion(operacion_elegida,casilla_elegida)
        end
        break if 1!=1
      end
    end
    
  end
  VistaTextualQytetet.main
end

File: Qytetet Java/src/vistatextualqytetet/VistaTextualQytetet.java
/*
 * To change this license header, choose License Headers in Project Properties.
 * To change this template file, choose Tools | Templates
 * and open the template in the editor.
 */
package vistatextualqytetet;
import controladorqytetet.ControladorQytetet;
import controladorqytetet.OpcionMenu;
import java.util.ArrayList;
import java.util.Scanner;

/**
 *
 * @author migue
 */
public class VistaTextualQytetet {
    private static ControladorQytetet controlador = ControladorQytetet.Instance();
    private static final Scanner in =new Scanner (System.in);
    
    public ArrayList<String> obtenerNombreJugadores(){
        int numeroJugadores;
        ArrayList<String> nombres = new ArrayList<>(); 
        System.out.println("Introduzca el numero de jugadores (entre 2 y 4)");{
            do{
                numeroJugadores = in.nextInt();
            }while (numeroJugadores < 2 && numeroJugadores > 4);
        }
        for (int i=0; i < numeroJugadores ;i++){
            System.out.println("Introduzca el nombre del jugador "+i);
            String s = in.next();
            nombres.add(s);
        }
          return nombres;  
    }
    public int elegirCasilla(int opcionMenu){
        ArrayList<Integer> lista=controlador.obtenerCasillasValidas(opcionMenu);
        ArrayList<String> casillas = new ArrayList();
        if(lista.isEmpty())
            return -1;
        else{
            System.out.println("Las casillas válidas son: ");
            for(int i=0;i<lista.size();i++){
                System.out.println(lista.get(i));
                casillas.add(Integer.toString(lista.get(i)));
            }
        }
            
            return Integer.parseInt(leerValorCorrecto(casillas,false));
        
    }
    public String leerValorCorrecto(ArrayList<String> valoresCorrectos, boolean opcion){
        String valorLeido, valorRetornado=null;
        boolean correcto=false;
        System.out.println("Selecciona una de las siguientes opciones: ");
        if(opcion){
            for(int i=0;i<valoresCorrectos.size();i++){
                System.out.println("Opción: "+valoresCorrectos.get(i) + " - " + OpcionMenu.values()[Integer.parseInt(valoresCorrectos.get(i))]);
             }
        }
        else{
             for(int i=0;i<valoresCorrectos.size();i++){
                System.out.println("Casilla :" + valoresCorrectos.get(i));
             }
        }

        do{
            valorLeido=in.next();
            for(int i=0;i<valoresCorrectos.size();i++){
                if(Integer.parseInt(valorLeido)==Integer.parseInt(valoresCorrectos.get(i))){
                    correcto=true;
                    valorRetornado=valorLeido;
                }
            }
        }while(!correcto);
        return valorRetornado;
    }
    
    public int elegirOperacion(){
        ArrayList<Integer> lista = controlador.obtenerOperacionesJuegoValidas();
        ArrayList<String> operaciones=new ArrayList();
        
        for(int i=0;i<lista.size();i++){
            operaciones.add(Integer.toString(lista.get(i)));
        }
        return Integer.parseInt(this.leerValorCorrecto(operaciones,true));
    }
    
    public static void main(String[] args) {
        VistaTextualQytetet ui = new VistaTextualQytetet();
        controlador.setNombreJugadores(ui.obtenerNombreJugadores());
        int operacionElegida, casillaElegida=0;
        boolean necesitaElegirCasilla;
        
        do{
            operacionElegida=ui.elegirOperacion();
            necesitaElegirCasilla=controlador.necesitaElegirCasilla(operacionElegida);
            if(necesitaElegirCasilla)
                casillaElegida=ui.elegirCasilla(operacionElegida);
            if(!necesitaElegirCasilla || casillaElegida>=0)
                System.out.println(controlador.realizarOperacion(operacionElegida, casillaElegida));
            
        }while(1==1);
    }
}
